app/controllers/visualizations_controller.rb

```diff
@@ -126,9 +126,67 @@ def shorten_url
     }, status: :internal_server_error
   end
 
+  # START OF HEATMAP METHODS
+
+  def heatmap
+    {
+      taxonLevels: %w[Genus Species],
+      categories: ReportHelper::ALL_CATEGORIES.pluck('name'),
+      subcategories: {
+        Viruses: ["Phage"]
+      },
+      metrics: [
+        { text: "NT Z Score", value: "NT.zscore" },
+        { text: "NT rPM", value: "NT.rpm" },
+        { text: "NT r (total reads)", value: "NT.r" },
+        { text: "NR Z Score", value: "NR.zscore" },
+        { text: "NR r (total reads)", value: "NR.r" },
+        { text: "NR rPM", value: "NR.rpm" }
+      ],
+      backgrounds: current_power.backgrounds.map do |background|
+        { name: background.name, value: background.id }
+      end,
+      thresholdFilters: {
+        targets: [
+          { text: "NT Z Score", value: "NT_zscore" },
+          { text: "NT rPM", value: "NT_rpm" },
+          { text: "NT r (total reads)", value: "NT_r" },
+          { text: "NT %id", value: "NT_percentidentity" },
+          { text: "NT L (alignment length in bp)", value: "NT_alignmentlength" },
+          { text: "NT log(1/e)", value: "NT_neglogevalue" },
+          { text: "NR Z Score", value: "NR_zscore" },
+          { text: "NR r (total reads)", value: "NR_r" },
+          { text: "NR rPM", value: "NR_rpm" },
+          { text: "NR %id", value: "NR_percentidentity" },
+          { text: "NR L (alignment length in bp)", value: "NR_alignmentlength" },
+          { text: "R log(1/e)", value: "NR_neglogevalue" }
+        ],
+        operators: [">=", "<="]
+      },
+      allowedFeatures: current_user.allowed_feature_list
+    }
+  end
+
+  def download_heatmap
+    @sample_taxons_dict = HeatmapHelper.sample_taxons_dict(params, samples_for_heatmap)
+    output_csv = generate_heatmap_csv(@sample_taxons_dict)
+    send_data output_csv, filename: 'heatmap.csv'
+  end
+
+  def samples_taxons
+    @sample_taxons_dict = HeatmapHelper.sample_taxons_dict(params, samples_for_heatmap)
+    render json: @sample_taxons_dict
+  end
+
   private
 
   def visualization_params
     params.permit(:domain, :type, :id, :url, :search, data: {})
   end
+
+  def samples_for_heatmap
+    current_power.samples
```

Comment:
Since we are moving, this looks like a model kind of function.

Reply:
isn't `current_power` required to be in a controller because it uses session state?

Reply:
Yeah, nevermind, I guess the key here for samples is the includes statement which is heatmap specific. 
I would just see the rest as something like `current_power.visualization.authorized_samples` which you would define through power controller. 

In any case, disregard this comment on the context of this PR.